### PR TITLE
schnorr signature draft

### PR DESCRIPTION
**Pull request by odyssey2077**

### CODE REVIEW

File: halo2-ecc/src/secp256k1/tests/schnorr_signature_tests.rs [hunk 1]
```diff
@@ -0,0 +1,349 @@
+#![allow(non_snake_case)]
```
Comment (odyssey2077) [inline halo2-ecc/src/secp256k1/tests/schnorr_signature_tests.rs:1]: all tests pass locally, didn't update the benchmark csv for schnorr

File: halo2-base/src/gates/flex_gate.rs [hunk 2]
```diff
@@ -787,6 +789,35 @@ pub trait GateInstructions<F: ScalarField> {
...
+    fn is_even(&self, ctx: &mut Context<F>, a: AssignedValue<F>) -> AssignedValue<F> {
+        let x = fe_to_biguint(a.value());
+        let (div, rem) = x.div_mod_floor(&BigUint::from(2u64));
+
+        let one_minus_rem = biguint_to_fe(&(BigUint::from(1u64) - &rem));
+        let div = biguint_to_fe(&div);
+        let rem = biguint_to_fe(&rem);
+
+        let cells = [
+            Witness(rem),
```
Comment (jonathanpwang) [inline halo2-base/src/gates/flex_gate.rs]: since you aren't using any advanced overlapping in `assign_region` (offsets are multiples of 4), I recommend trying to use the existing `GateInstructions` functions as much as possible to improve readability and security

File: halo2-base/src/gates/flex_gate.rs [hunk 3]
```diff
@@ -787,6 +789,35 @@ pub trait GateInstructions<F: ScalarField> {
...
 
+    /// Constrains that a cell value is even and returns `1` if `a is even`, otherwise `0`.
+    /// * `ctx`: [Context] to add the constraints to
+    /// * `a`: [QuantumCell] value to be constrained
+    fn is_even(&self, ctx: &mut Context<F>, a: AssignedValue<F>) -> AssignedValue<F> {
+        let x = fe_to_biguint(a.value());
+        let (div, rem) = x.div_mod_floor(&BigUint::from(2u64));
+
+        let one_minus_rem = biguint_to_fe(&(BigUint::from(1u64) - &rem));
+        let div = biguint_to_fe(&div);
```
Comment (jonathanpwang) [inline halo2-base/src/gates/flex_gate.rs]: `div` needs to be range checked: you can use `RangeInstructions::get_last_bit` function as a stand-in for `is_even`

File: halo2-ecc/src/ecc/schnorr_signature.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    GA: CurveAffineExt<Base = CF, ScalarExt = SF>,
+{
+    let base_chip = chip.field_chip;
+    let scalar_chip =
+        FpChip::<F, SF>::new(base_chip.range, base_chip.limb_bits, base_chip.num_limbs);
+
+    // check r < p
+    let r_valid = base_chip.is_soft_nonzero(ctx, &r);
+    // check s < n
+    let s_valid = scalar_chip.is_soft_nonzero(ctx, &s);
```
Comment (jonathanpwang) [inline halo2-ecc/src/ecc/schnorr_signature.rs:35]: this prevents `s = 0`. I'm not sure if that is an edge case

File: halo2-ecc/src/ecc/schnorr_signature.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+) -> AssignedValue<F>
+where
+    GA: CurveAffineExt<Base = CF, ScalarExt = SF>,
+{
+    let base_chip = chip.field_chip;
+    let scalar_chip =
+        FpChip::<F, SF>::new(base_chip.range, base_chip.limb_bits, base_chip.num_limbs);
+
+    // check r < p
+    let r_valid = base_chip.is_soft_nonzero(ctx, &r);
```
Comment (jonathanpwang) [inline halo2-ecc/src/ecc/schnorr_signature.rs]: This prevents `r = 0`. For most curves there's no way `x(R) = 0` so it seems technically ok, but wanted to check if that's the intention

File: halo2-ecc/src/ecc/schnorr_signature.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    );
+
+    // check s_G.x != e_P.x, which is a requirement for sub_unequal
+    let x_eq = base_chip.is_equal(ctx, &s_G.x, &e_P.x);
+    let x_neq = base_chip.gate().not(ctx, x_eq);
+
+    // R = s⋅G - e⋅P
+    let R = chip.sub_unequal(ctx, s_G, e_P, true);
+
+    // how to check if R is at inifinity? infinity means y = 0 && x == 0
```
Comment (jonathanpwang) [inline halo2-ecc/src/ecc/schnorr_signature.rs]: `R` infinity is the same as `s_G == e_P`. Since you did `sub_unequal` with strict = true, the circuit will always impose that `s_G != e_P`

File: halo2-ecc/src/ecc/schnorr_signature.rs (hunk 4 above)
Comment (jonathanpwang) [reply halo2-ecc/src/ecc/schnorr_signature.rs:35]: I guess `fixed_base::scalar_multiply` cannot handle `s = 0` right now anyways

File: halo2-ecc/src/secp256k1/tests/schnorr_signature_tests.rs [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+    let pk = ecc_chip.assign_point(ctx, pk);
+    // test schnorr signature
+    let res = schnorr_verify_no_pubkey_check::<F, Fp, Fq, Secp256k1Affine>(
+        &ecc_chip, ctx, pk, r, s, m, 4, 4,
+    );
+    assert_eq!(res.value(), &F::one());
+}
+
+fn random_parameters_schnorr_signature() -> (Fp, Fq, Fq, Secp256k1Affine) {
+    let sk = <Secp256k1Affine as CurveAffine>::ScalarExt::random(OsRng);
```
Comment (jonathanpwang) [inline halo2-ecc/src/secp256k1/tests/schnorr_signature_tests.rs]: use seeded rng for reproducibility

**Review by jonathanpwang — COMMENTED** [review]
Might be slightly over-constrained due to point at infinity edge cases, but mostly due to some incomplete-ness of support for point at infinity in halo2-ecc.

Also in cases of a failed signature, the circuit may in some cases have constraint failure instead of returning `0`. Since the main usage will just be constraining the output is `1`, this seems not a huge deal.

File: halo2-ecc/src/ecc/schnorr_signature.rs (hunk 5 above)
Comment (odyssey2077) [reply halo2-ecc/src/ecc/schnorr_signature.rs]: updated to not prevent r = 0

File: halo2-base/src/gates/flex_gate.rs (hunk 2 above)
Comment (odyssey2077) [reply halo2-base/src/gates/flex_gate.rs]: switched to using get_last_bit
